### ISSUE TITLE
Could this be a PMPI library?

### ISSUE THREAD

**Issue by tgamblin** [rookiehpc/MPI_monitor#1]
From the `README`:

> This library has been designed to be very quick and convenient to use:
> 
> 1. Replace all your `#include <mpi.h>` with `#include "mpi_monitor.h"`
> 2. Add the `mpi_monitor.c` file to your compilation command
> 3. Run your application as usual

Step (1) can be difficult for a large-ish code -- there are likely many files that include `mpi.h`.  Also, the instructions only support C programs -- you won't be able to use this with Fortran (or, [mpi4py](https://mpi4py.readthedocs.io/en/stable/)) programs as implemented.

It looks like `mpi_monitor.h` name-shifts all the `MPI_*` routines to `MPIM_*` routines.  But there is already an interface built into MPI to support this kind of thing: [PMPI](https://www.open-mpi.org/faq/?category=perftools#PMPI).  All (conforming) MPI implementations are required to have `PMPI_Foo` for every `MPI_Foo`, and this allows you to easily write *libraries* that can intercept any MPI call (as `MPI_monitor` does), do whatever the tool needs to do, and call the `PMPI_Foo` routine underneath to do the real work.

If this were implemented as a PMPI library (e.g., `libmpimonitor.so`), a user could intercept MPI calls in already-compiled MPI programs like this:

```console
export LD_PRELOAD=/path/to/libmpimonitor.so
mpirun -n 4 myprogram
```

or they could compile against the `libmpimonitor.so` library (if they wanted the tool baked in).

You can also handle Fortran programs "for free" if you use tools like https://github.com/LLNL/wrap to generate the interceptor library.  `wrap` makes it easy to intercept a lot of routines, and it handles the tricky aspects of MPI_Init in Fortran (and things like argument marshaling so you can write your tool's code in C but intercept Fortran calls).

Could `MPI_monitor` be implemented as a PMPI tool?

**Comment by rookiehpc** [comment]
> Step (1) can be difficult for a large-ish code -- there are likely many files that include mpi.h.

In the educational context, which is the primary focus, codes typically consist of one or two files. However, in practice, you are right, larger codes may require to replace that one line in multiple locations indeed, although IDEs or a single `sed` command certainly greatly reduce the time needed to accomplish that nowadays.

> It looks like mpi_monitor.h name-shifts all the MPI_* routines to MPIM_* routines. But there is already an interface built into MPI to support this kind of thing: [PMPI](https://www.open-mpi.org/faq/?category=perftools#PMPI). All (conforming) MPI implementations are required to have PMPI_Foo for every MPI_Foo, and this allows you to easily write libraries that can intercept any MPI call (as MPI_monitor does), do whatever the tool needs to do, and call the PMPI_Foo routine underneath to do the real work.

You are right, what those defines do includes name shifting. Although maybe not for the reasons one may think of at first. The MPI monitor does not sit between the original MPI interface and the PMPI one, but rather between the user and the original MPI interface. This is due to the need of extra information that must be fetched initially, as behind the scenes it fetches the line number, function name and file name of the originating location of the function call. As you can see, the signatures of the `MPIM_` functions typically end with these three arguments, thus differ from `MPI_` and `PMPI_` signatures. Fetching these information from within a customised `MPI_` function would result in that information to be local to the `MPI_` function, whereas the objective is to have the information from the place from which that function was called originally.

(Also, not to rely on PMPI avoids potential conflict with an existing PMPI library in place. The idea was really to have a tool guaranteed to work simply by appending `_monitor` to a single-line of code and recompile.)

I think it is the need for the extra information that makes the MPI monitor incompatible with the more conventional PMPI approach, as it stand at a different abstraction layer. What do you think?

**Comment by tgamblin** [comment]
> I think it is the need for the extra information that makes the MPI monitor incompatible with the more conventional PMPI approach, as it stand at a different abstraction layer. What do you think?

Yep, that's true -- you can't get `__FILE__`, `__LINE__`, `__FUNCTION__` etc. at runtime -- they're only available in the at compile time.

What we typically do in PMPI tools is use call stack unwinding to get the context for measurements, so, e.g. [this library](https://github.com/llnl/callpath), which sits on top of [GNU `backtrace`](https://www.gnu.org/software/libc/manual/html_node/Backtraces.html) or [dyninst](https://github.com/dyninst/dyninst).  can get you that information at runtime.  Stack unwinding is more complicated than what you're doing but it would give you more context -- e.g. if the same function is called from many places, you could see that. It may also add more overhead to your program if it's called from some very frequently invoked MPI routine (like `MPI_Test` in a loop) but if you're viewing the calls on the terminal that likely isn't important.

So, anyway, yes you *can* get some of that information this way but it's more work than what you're doing here.

I think this *could* have advantages in an educational context -- people usually just write mpirun `wrappers` for tools like this, so you can invoke the tool like this:

```console
monitor-mpirun -n 4 myprogram
```

But if the user has to deal with building the tool, it could end up being more complex for them than what you have.

**Comment by rookiehpc** [comment]
> But if the user has to deal with building the tool, it could end up being more complex for them than what you have.

This was actually the original motivation for this design: more often than not, a tool will need multiple passes to correctly build. The first version of this MPI monitor was header-only, to leave the compilation phase unmodified. However, it was limiting the target code to a single file, otherwise raising a multiple redefinitions error as expected. Nonetheless, limiting the user code to a single-file was too restrictive, even for toy examples which can use a main file with one or two additional files, so the MPI monitor was redesigned to work in multi-file contexts. It requires an additional compilation, which is kept convenient as it consists of a single file.

> I think this could have advantages in an educational context -- people usually just write mpirun wrappers for tools like this, so you can invoke the tool like this:
```
monitor-mpirun -n 4 myprogram
```

From the feedback received, it seems that one of the main difficulties when learning MPI is distinguishing `mpicc` from being a compiler, a script or a library. The notion of wrapper tends to be confusing to MPI beginners, especially since it seems there is a lot of _magic_ already in the `mpi*` commands. Therefore, wrapping around wrappers or scripts is not a technique I tend to actively explore.

Of course, the above no longer holds when considering learners with more experience, who are comfortable relying on more complex tools. However, I assume that intermediate or advanced users will go for "real" tools providing a much deeper insight. The MPI monitor is really centred around beginners learning `MPI_Send`/`MPI_Recv` and encountering their first deadlock when using a collective operation.

However, should you be interested in expanding the MPI monitor into a new tool by making it more complex and able, I would be happy to discuss about that project too.
